File: apps/ios/Pebbles/Features/Profile/Models/SoulWithGlyph.swift
import Foundation

/// A soul together with its joined glyph and the live count of pebbles
/// linked to it. Decoded from a single PostgREST request:
///
///     supabase.from("souls")
///         .select("""
///             id, name, glyph_id,
///             glyphs(id, name, strokes, view_box),
///             pebbles_count:pebble_souls(count)
///         """)
///
/// PostgREST nests the joined glyph row under the relation name (`glyphs`)
/// and returns the aggregate as a single-element array of `{count: Int}`
/// objects (e.g. `[{ "count": 12 }]`). A soul with zero linked pebbles is
/// expected to come back as `[{ "count": 0 }]`; an empty array is treated
/// as 0 defensively.
struct SoulWithGlyph: Identifiable, Decodable, Hashable {
    let id: UUID
    let name: String
    let glyphId: UUID
    let glyph: Glyph
    let pebblesCount: Int

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case glyphId      = "glyph_id"
        case glyph        = "glyphs"
        case pebblesCount = "pebbles_count"
    }

    private struct CountWrapper: Decodable { let count: Int }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id      = try c.decode(UUID.self,   forKey: .id)
        name    = try c.decode(String.self, forKey: .name)
        glyphId = try c.decode(UUID.self,   forKey: .glyphId)
        glyph   = try c.decode(Glyph.self,  forKey: .glyph)
        let wraps = try c.decodeIfPresent([CountWrapper].self, forKey: .pebblesCount) ?? []
        pebblesCount = wraps.first?.count ?? 0
    }

    // Required: providing any initializer in the struct body suppresses
    // the synthesized memberwise init. Preview and test code construct
    // instances without a network response, so re-declare it explicitly.
    init(id: UUID, name: String, glyphId: UUID, glyph: Glyph, pebblesCount: Int) {
        self.id = id
        self.name = name
        self.glyphId = glyphId
        self.glyph = glyph
        self.pebblesCount = pebblesCount
    }

    /// Convenience for code paths that already hold a `Soul` and need to
    /// drop the joined glyph (e.g. passing into `EditSoulSheet` which only
    /// needs the `Soul` shape).
    var soul: Soul {
        Soul(id: id, name: name, glyphId: glyphId)
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/profile/models/SoulWithGlyph.kt
package app.pbbls.android.features.profile.models

import app.pbbls.android.features.glyph.models.Glyph
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Clean domain type — a soul + its glyph + live pebble count. Mirrors iOS SoulWithGlyph. */
data class SoulWithGlyph(
    val id: String,
    val name: String,
    val glyphId: String,
    val glyph: Glyph,
    val pebblesCount: Int = 0,
)

/**
 * Wire row for `souls` selects. PostgREST nests the joined glyph under the FK
 * relation name `glyphs` and returns `pebble_souls(count)` as `[{ "count": N }]`;
 * absent (the reference-data select doesn't ask for it) decodes to 0.
 */
@Serializable
data class SoulRow(
    val id: String,
    val name: String,
    @SerialName("glyph_id")
    val glyphId: String,
    @SerialName("glyphs")
    val glyph: Glyph,
    @SerialName("pebbles_count")
    val pebblesCount: List<CountRow> = emptyList(),
) {
    fun toSoulWithGlyph(): SoulWithGlyph =
        SoulWithGlyph(
            id = id,
            name = name,
            glyphId = glyphId,
            glyph = glyph,
            pebblesCount = pebblesCount.firstOrNull()?.count ?: 0,
        )

    @Serializable
    data class CountRow(
        val count: Int,
    )
}
